code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -527,47 +577,31 @@ runBroker = do
     -- Runner
     let withWorkers action =
              A.withAsync (heartbeatWorker ztCliSettings ztHeartbeatInfo ztCliRequestQueue) $ const $
+             A.withAsync (clientsToBrokerWorker cEnv) $ const $
              action
 
     liftIO $ withWorkers $ do
       let action = do
-              results <- atomically $ do
-                  cMap <- readTVar ztClients
-                  peersInfo <- HMap.toList <$> readTVar ztPeers
-                  let boolToMaybe t = bool Nothing (Just t)
-                  let readReq =
-                          fmap CBRequest <$>
-                          TQ.tryReadTQueue (unCliRequestQueue ztCliRequestQueue)
-                  let readClient :: [STM (Maybe CliBrokerStmRes)]
-                      readClient =
-                          map (\(cliId,biq) ->
-                                  fmap (\(nodeIdM,content) -> CBClient cliId nodeIdM content) <$>
-                                  TQ.tryReadTQueue (bSendQ biq))
-                              (Map.toList cMap)
-                  let readBacksSubs =
-                          concatMap
-                          (\(nId, PeerRes{..}) ->
-                            [ boolToMaybe (CBSub prSub nId) <$> adapterTry prSubAdapter
-                            , boolToMaybe (CBBack prBack nId) <$> adapterTry prBackAdapter])
-                          peersInfo
-                  atLeastOne $ NE.fromList $ readReq : readBacksSubs ++ readClient
-              forM_ results $ \case
-                  CBRequest r             -> processReq r
-                  CBClient _cId nIdM cont -> clientToBackend nIdM cont
-                  (CBBack dealer nId)     -> receiveBack dealer nId
-                  (CBSub sub nId)         -> receiveSub sub nId
-
-      -- DSCP-177 Sockets must be processed at least once in the beginning
-      -- for 'threadWaitRead' to function correctly later. This is not a
-      -- solution, but a workaround -- I haven't managed to find the
-      -- explanation of why does it happen (@volhovm).
-      -- TODO
-      -- TODO
-      -- TODO
-      -- TODO
-      -- TODO
-      --receiveBack
-      --receiveSub
+              CliPollData{..} <- atomically $ readTVar ztCliPollData
+
+              let resolve i =
+                      fromMaybe (error $ "lookupUnsafe " <> show (i,cpdN)) $
+                      HMap.lookup i cpdResolveMap
+
+              events <- Z.poll (-1) cpdPolls
+              forM_ (events `zip` [0..]) $ \(e,i) ->
+                  unless (null e) $
+                      if | i < cpdN ->
```

Comment:
honestly, I'd prefer to have separate function `classifySocketIdx` which returns the tag by socket index

Reply:
That's definitely not a healthy urge.

Reply:
lool, look at this code: some insane indexes which only you understand and only now, it's easier to find a bug in one function then in 4 compares among a lot of another code

Reply:
ne ne, it's not resolved

Reply:
I can do it myself in neighbour PR, but I insist it has to be done, because these checks
```
if | i < cpdN ->
| i < 2 * cpdN ->
| i == 2 * cpdN ->
i == 2 * cpdN + 1
```
really hard to follow, and I prefer to have it this way
```
case classifySockIndex i cpdN where
    BackSock peerId ->
    SubSock peerId   ->
    CliSock                 ->
    ReqSock               ->
```

Reply:
so if you are interested how I recognise what each brunch does: "taaks ebana, chto tut u nas v tele, aga `receiveSub`, pohodu eto sub socket", so until I look at body I can't recognise the case.